indexd_ffi/src/lib.rs

```diff
@@ -1133,20 +1271,29 @@ impl SDK {
 
     /// Retrieves a shared object from a signed URL.
     pub async fn shared_object(&self, shared_url: &str) -> Result<SharedObject, Error> {
-        let shared_object = self
-            .app_client
-            .shared_object(
-                shared_url
-                    .parse()
-                    .map_err(|e| Error::Custom(format!("{e}")))?,
-            )
-            .await?;
-        Ok(SharedObject(shared_object))
+        if !self.connected().await {
+            return Err(Error::NotConnected);
+        }
+        let app_client = self.app_client.clone();
+        let shared_url: Url = shared_url
+            .parse()
+            .map_err(|e| Error::Custom(format!("{e}")))?;
+        self.runtime
+            .spawn(async move {
+                let shared_object = app_client.shared_object(shared_url).await?;
+                Ok(SharedObject(shared_object))
+            })
+            .await?
     }
 
     /// Pins a shared object to the indexer and returns a [PinnedObject].
     pub async fn pin_shared(&self, shared: Arc<SharedObject>) -> Result<PinnedObject, Error> {
+        if !self.connected().await {
```

Comment:
Not sure if there is a good way to do it but looking at this really makes me wish we could lock `self.app_client` behind `connected`. So that you always need to call `connected` to obtain the client.

Reply:
Type states + UniFFI don't mix well. Since we can't consume the locked type, this is a more understandable DX. This works fine.

Reply:
Gave it a go since I agree it's kind of lame. On the fence on whether it's nicer for a new developer or not. https://github.com/SiaFoundation/sia-sdk-rs/pull/194